File: 208-ImplementTriePrefixTree/208-ImplementTriePrefixTree.py
# Last updated: 6/11/2026, 11:34:57 AM
class TrieNode:
    def __init__(self):
        self.children = {} #can be dictionary of all 26 letters or an array of 26 letters
        self.endofword = False

class Trie:
    def __init__(self):
        self.head = TrieNode()
    def insert(self, word: str) -> None:
        cur = self.head
        for c in word:
            if c in cur.children:
                cur = cur.children[c]
            else:
                node = TrieNode()
                cur.children[c] = node
                cur = node # or cur = cur.children[c]
        cur.endofword = True

    def search(self, word: str) -> bool:
        cur = self.head
        for c in word:
            if c not in cur.children:
                return False
            cur = cur.children[c]
        return cur.endofword

    def startsWith(self, prefix: str) -> bool:
        cur = self.head
        for c in prefix:
            if c in cur.children:
                cur = cur.children[c]
            else:
                return False
        return True


# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)

File: 208-ImplementTriePrefixTree/208-ImplementTriePrefixTree.cpp
// Last updated: 6/11/2026, 11:34:52 AM
struct Node{
    vector<Node*> arr;
    int ends = 0;
    Node(){
        arr.resize(26, nullptr);
        ends = 0;
    }
};


class Trie {
public:
    Node* root;
    Trie() {
        root = new Node();
    }
    
    void insert(string word) {
        Node* cur = root;
        for(char ch: word){
            if(cur->arr[ch-'a']==nullptr){
                cur->arr[ch-'a'] = new Node();
            }
            cur = cur->arr[ch-'a'];
        }
        cur->ends++;
    }
    
    bool search(string word) {
        Node* cur = root;
        for(char ch: word){
            if(cur->arr[ch-'a'] == nullptr) return 0;
            cur = cur->arr[ch-'a'];
        }
        return cur->ends>0;
    }
    
    bool startsWith(string prefix) {
        Node* cur = root;
        for(char ch: prefix){
            if(cur->arr[ch-'a'] == nullptr) return 0;
            cur = cur->arr[ch-'a'];
        }
        return 1;
    }
};

/**
 * Your Trie object will be instantiated and called as such:
 * Trie* obj = new Trie();
 * obj->insert(word);
 * bool param_2 = obj->search(word);
 * bool param_3 = obj->startsWith(prefix);
 */